draft-ietf-dnsop-svcb-https.md

```diff
@@ -1082,18 +1082,58 @@ single-client granularity.
 
 ## Interaction with Alt-Svc
 
-Clients that do not implement support for Encrypted ClientHello MAY
-skip the HTTPS RR query
-if a usable Alt-Svc value is available in the local cache.
-If Alt-Svc connection fails, these clients SHOULD fall back to the HTTPS RR
-client connection procedure ({{client-behavior}}).
-
-Clients that implement support for ECH MUST perform the HTTPS RR query first,
-and MUST only make use of Alt-Svc when operating in SVCB-optional mode (see
-{{ech-client-behavior}}).
-
-This specification does not alter the DNS records used when connecting
-to an Alt-Svc hostname (typically A and/or AAAA only).
+Clients that implement support for both Alt-Svc and HTTPS records SHOULD
+retrieve any HTTPS records for the Alt-Svc alt-authority, and ensure that
+their connection attempts are consistent with both the Alt-Svc parameters
+and any received HTTPS SvcParams.  If present, the HTTPS record's TargetName
+and port override the alt-authority.  For example, suppose that
+"https://example.com" sends an Alt-Svc field value of:
+
+~~~ HTTP
+Alt-Svc: h2="alt.example:443", h2="alt2.example:443", h3=":8443"
+~~~
+
+The client would retrieve the following HTTPS records:
+
+    alt.example.              IN HTTPS 1 . alpn=h2,h3 ech=...
+    alt2.example.             IN HTTPS 1 alt2b.example. alpn=h3 ech=...
+    _8443._https.example.com. IN HTTPS 1 alt3.example. (
+        port=9443 alpn=h2,h3 ech=... )
+
+Based on these inputs, the following connection attempts would always be
+allowed:
+
+* HTTPS over TLS to `alt.example:443` (Alt-Svc and HTTPS record)
+* HTTP/3 to `alt3.example:9443` (Alt-Svc and HTTPS record)
+* Fallback to the the client's non-Alt-Svc connection behavior
+
+ECH-capable clients would use ECH when establishing any of these connections.
+
+The following connection attempts would not be allowed:
+
+* HTTP/3 to `alt.example:443` (not consistent with Alt-Svc)
+* Any connection to `alt2b.example` (no ALPN consistent with both the HTTPS
+  record and Alt-Svc)
+* HTTPS over TLS to any port on `alt3.example` (not consistent with Alt-Svc)
+
+The following connection attempts would be allowed only if the client does
+not support ECH, as they rely on SVCB-optional fallback behavior that is
+disabled when the "ech" SvcParam is present ({{ech-client-behavior}}):
+
+* HTTPS over TLS to `alt2.example:443` (Alt-Svc only)
+* HTTP/3 to `example.com:8443` (Alt-Svc only)
+
+Origins that publish an "ech" SvcParam in their HTTPS record SHOULD
+also publish an "ech" SvcParam for any alt-authorities.  Otherwise,
+clients might reveal the name of the server in an unencrypted ClientHello.
+Similar consistency considerations could apply to future SvcParamKeys, so
+alt-authorities SHOULD carry the same SvcParams as the origin unless
+a deviation is specifically known to be safe.
+
```

Comment:
Should we point out that there may be areas of transition where there may be inconsistencies?  For example, when origins are adding ECH?  (Due to the lack of an SVCB-Used or equivalent, the server returning the Alt-Svc has no way to know which version of an HTTPS RR was returned.  As such, there is may be mismatch during transitions.  This is probably OK, but clients MUST NOT break when there is a mismatch.

Reply:
Clients would have to go out of their way to create brokenness here, so I don't think we need this warning.

Reply:
As long as we're just talking about Alt-Svc, I think it's acceptable for a client to completely "break" and disallow the entire Alt-Svc for whatever reason the client chooses.  Just the nature of Alt-Svc as an optional mechanism.

Reply:
Ignoring the Alt-Svc entirely in this case is fine, as long as the client falls back to the standard origin.  I guess that is clear in Alt-Svc so we don't need to spell it out here.